### ISSUE TITLE
Pi-hole should not install resolvconf and hi-jack local DNS resolving by default

### ISSUE THREAD

**Issue by hkbakke** [pi-hole/pi-hole#2988]
Pi-hole used an outdated approach to local DNS resolver configuration that is not really in tune with the way things are handled in modern distributions using systemd-resolved running on 127.0.0.53 as their local resolver.

The first issue is that resolvconf is installed, and it immediately starts to mess with local DNS settings in /etc/resolv.conf, mainly the reasoning being that Pi-hole want to point to itself as the local resolver. This is quite unacceptable default behaviour for a DNS _server_, as local DNS-resolver settings really has nothing to do with the service installed, which may or may not work as intended, breaking local resolving that might be completely differently set up.

Might I suggest that the default install configuration is for a setup not messing with local DNS resolver settings? There are many ways to configure and handle local DNS resolving, and the way Pi-hole works by default today is really only set up for a single assumption that it is used on a desktop client, which in my opinion is plain unnatural to have as the default setting. 

It is quite easy to change local resolver to 127.0.0.1, either in resolv.conf, or in modern distros, in /etc/systemd/resolved.conf, or even in dhclient.conf, if that is what one want, but it is much harder to work against Pi-hole continuously trying it's best to mess everything up on every new installs (which reinstalls resolvconf).

Also, to work nicely with systemd-resolved by default, you should concider adding "bind-interfaces" to the dnsmasq config by default when "only bind to selected interfaces" is selected as the preferred config mode. This ensures that systemd-resolved can own 127.0.0.53 port 53 for local resolver cache while Pi-hole binds only to the specified interfaces.

To summarize:
* It is unexpected that Pi-hole touches resolv.conf and adds itself, so don't
* Don't install resolvconf at all. If the system owner wants to use that specific method for DNS configuration it should be his/her own choice.
* Add `bind-interface` to dnsmasq config if "Listen only on interface" is chosen as the preferred mode in Pi-hole

**Comment by dschaper** [comment]
This comes up in discussion quite often. The counter to not using Pi-hole as the resolver on the DNS server is for people using the Raspberry Pi as a workstation. We've talked a lot about this as a team internally but we haven't made a firm decision to change the current design.

Personally I agree, we already have to work around the possibility that the Pi-hole blocks itself from being able to source and update lists. I'd like to have @PromoFaux @DL6ER and @bcambl to chip in though as they may see things that I haven't.

**Comment by hkbakke** [comment]
It is possible to keep the behaviour of today (although I don't think its the proper default myself), but ask  "Do you want to use Pi-hole as this hosts DNS resolver" action during installation in the same way that other configuration questions are handled. This option must stick between upgrades. That way most people should be happy.

**Comment by dschaper** [comment]
Any new dialog boxes or questions just get clicked right past and the default is used. I'm not against a choice but what ever the default answer is is the one that will be used most of the time.

**Comment by bcambl** [comment]
My thoughts: https://github.com/pi-hole/pi-hole/pull/2504#discussion_r240010778

**Comment by dschaper** [comment]
https://github.com/pi-hole/pi-hole/issues/2535

**Comment by festro** [comment]
I'm wondering if this is the issue I'm having as well? I set up the pihole and it refuses to update gravity, the ubuntu server I run it on (apt update) it just outright refuses the connection. The workaround I had been using when this occurs is this command " echo "nameserver 1.1.1.1" | sudo tee /etc/resolv.conf > /dev/null" but running this to update the system or pihole, and then reconfiguring it every time is pretty cumbersome.

**Comment by darrenallendunn** [comment]
I second the OP’s comments regarding PiHole destroying DNS configuration. I’m currently running Ubuntu 18.04 LTS server as well as Elementary OS (built upon LTS) and either can’t connect to the net upon system boot (DNS service not running / The FTL service is offline!) due to port 53 being in use or randomly lose connection. After crawling through discussions here, Discourse and Reddit, this seems to be an issue affecting lots of Ubuntu folks which is odd considering it’s one of the top five largest operating systems on the planet. Are there any direct, non hacked, solutions for this on the roadmap?

**Comment by DL6ER** [comment]
@darrenallendunn See the referenced issue ticket https://github.com/pi-hole/pi-hole/issues/2535

It is on the roadmap to leave the local DNS configuration of the machine you install Pi-hole on untouched. All clients that should use the Pi-hole need to be configured in one way or the other in this case (auto for DHCP, manually for static addresses). The server does not need to be special compared to the clients. @dschaper for confirmation?

**Comment by dschaper** [comment]
Yes, it's under consideration.

**Comment by MichaIng** [comment]
I vote as well for dropping resolvconf dependency. The init script already dropped calling it, if I see it right?
v4: https://github.com/pi-hole/pi-hole/blob/master/advanced/Templates/pihole-FTL.service#L51
v5: https://github.com/pi-hole/pi-hole/blob/development/advanced/Templates/pihole-FTL.service
Commit: https://github.com/pi-hole/pi-hole/commit/620e1e9c73277fb1f6c1f0be3a8cc3bd9d2b82f5#diff-045722c02800a1422bcac00099aec6db

Generally removing the need for localhost nameserver entry is a nice step, e.g. as it avoids possible conflicts between multiple local resolvers, i.e. `systemd-resolved` (default on Ubuntu) <> `resolvconf`. Also it makes sense that any client, including the server, is free to choose using Pi-hole as resolver or not. For the server it is from user perspective mostly not required, as long as the machine is not also used as desktop with browser or similar.

**Comment by PromoFaux** [comment]
Yep, already decided to drop it internally, I just have not got around to making the PR yet to actually drop it.

We also, I think, need to consider what we set as the nameserver in `dhcpcd.conf`, which currently is set to localhost. will it cause issues? I'm yet to test!

**Comment by dschaper** [comment]
No changes should be made to `dhcpcd.conf`, so I think we'll need to remove that line from the installer as well. (The one that writes 127.0.0.1 as the nameserver.)

**Comment by pralor** [comment]
This issue has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/pihole-doesnt-seem-to-be-getting-any-queries-can-still-see-ads-dns-seems-ok/28622/16

**Comment by PromoFaux** [comment]
So, where we currently do this:

```bash
        echo "interface ${PIHOLE_INTERFACE}
        static ip_address=${IPV4_ADDRESS}
        static routers=${IPv4gw}
        static domain_name_servers=127.0.0.1" | tee -a /etc/dhcpcd.conf >/dev/null
```

we should be doing this?

```bash
        echo "interface ${PIHOLE_INTERFACE}
        static ip_address=${IPV4_ADDRESS}
        static routers=${IPv4gw}  | tee -a /etc/dhcpcd.conf >/dev/null
```

**Comment by dschaper** [comment]
Oui oui monsieur

**Comment by MichaIng** [comment]
@PromoFaux 
Hmm, but a static IP setup requires a static DNS server entry as well, doesn't it? If there is ifupdown /etc/network/interfaces configured, it will setup the DNS server entry according to it's content, but if dhcpcd is used as only network tool, what does it do without an explicit nameserver entry and none assigned through DHCP of course?

**Comment by bcambl** [comment]
For rpm dists, I have always set the local nameservers to the upstreams chosen by the user via whiptail dialogs.
That’s why we typically only see the 127.0.0.1 issue on deb dists.

**Comment by dschaper** [comment]
Setting it as the upstream sounds logical. I think Raspbian works with the DHCP advertised DNS server even if the IP is set to static. Not in a place to be able to check for a few days on my end though.

**Comment by PromoFaux** [comment]
So, like this:

https://github.com/pi-hole/pi-hole/pull/3207/commits/175d32c5f660a03368be40bb931a3752bb24643c

I think I have the syntax right.

**Comment by MichaIng** [comment]
@bcambl 
Sounds like a good solution. If any static DNS has been chosen before, it is likely that users will configure the same as Pi-hole upstream DNS anyway. This could be even added as pre-defined entries/array to the installers DNS list, pasing /etc/resolv.conf nameserver entry/entries?
It is then assured as well that this DNS server is funktional, for Pi-hole, as well as for the server itself when Pi-hole is down.

Still, if there was DHCP configured with dhcpcd before, and then it is reconfigured to assign a static IP but no static DNS, does it still send DHCP requests to receive a DNS nameserver? Simple to test, will do later.

And just in case it is relevant:
- In many cases, when the router serves as DHCP server, it assigns itself as DNS nameserver as well.
- Now, when this is assigned as local DNS nameserver, and users configure their router to use Pi-hole as upstream DNS, then the Pi-hole machines DNS requests are basically in a loop?

**Comment by yubiuser** [comment]
Closing as https://github.com/pi-hole/pi-hole/commit/175d32c5f660a03368be40bb931a3752bb24643c is part of Pi-hole v5.0 which has been released long time ago. Pi-hole does not set itself as DNS server in `resolv.conf`
